### ISSUE TITLE
Add files MD5 hash to requests

### ISSUE THREAD

**Issue by jamestalmage** [github/git-lfs#722]
Google Cloud Storage (for one), will only validate MD5 hashes of the upload payload. They have no SHA-256 option.

It [looks like](http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html) S3 does support SHA-256 hashes of the payload, but not with pre-signed URLs. I have spent a long while trying to get the headers set correctly for that. So far I have been unsuccessful (if somebody has this working I would love pointers).

S3 will not validate the SHA-256 in [chunked uploads](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html) either.

Basically, what I keep running into is that having the client provide the MD5 would make implementing a server a LOT easier.

**Comment by heavenlyhash** [comment]
MD5 is flatly unsafe to use as an integrity mechanism: see http://www.mscs.dal.ca/~selinger/md5collision/ for reference and fully worked demonstration.

Regardless of whether or not GCS and AWS have joined us in the current century, "supporting" a hash with known collisions and readymade tools to freely generate more would be a ridiculously grave mistake for a content-addressable storage system.

It's time to let MD5 go.

**Comment by jamestalmage** [comment]
Some thoughts on that: 

  * Both services require SHA-256 signatures to authenticate requests.
  * I am talking about error-checking, not addressing or security.
  * You would still use the SHA-256 hash as the key/address of the file.
  * You would still have the client validate the SHA-256 hash after download.
  * MD5 seems perfectly adequate for file-revving and error-checking. Collisions require malicious behavior, and again - I am not recommending it as a security mechanism.

**Comment by technoweenie** [comment]
Hey all, I'm going through all the old issues that haven't been triaged, and replying to them.

Thanks for the request. We're not going to add MD5 hashes to requests, but we are kicking around some ideas that could lead to a GCS adapter in the future: https://github.com/github/git-lfs/compare/experimental/transfer-features#diff-2548f7e2abc2db33e7f0853da567dd1b
